src/libsystemd/sd-json/json-util.c

```diff
@@ -281,3 +282,63 @@ int json_dispatch_pidref(const char *name, sd_json_variant *variant, sd_json_dis
 
         return 0;
 }
+
+static int json_variant_new_stat(sd_json_variant **ret, struct stat *st) {
+        char mode[64];
+
+        assert(st);
+
+        if (!stat_is_set(st))
+                return sd_json_variant_new_null(ret);
+
+        xsprintf(mode, "%04o", st->st_mode & ~S_IFMT);
+
+        return sd_json_buildo(
+                        ret,
+                        SD_JSON_BUILD_PAIR(
+                                        "dev",
+                                        SD_JSON_BUILD_OBJECT(
+                                                        SD_JSON_BUILD_PAIR_UNSIGNED("major", major(st->st_dev)),
+                                                        SD_JSON_BUILD_PAIR_UNSIGNED("minor", minor(st->st_dev)))),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("inode", st->st_ino),
+                        SD_JSON_BUILD_PAIR_STRING("type", inode_type_to_string(st->st_mode)),
+                        SD_JSON_BUILD_PAIR_STRING("mode", mode),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("linkCount", st->st_nlink),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("uid", st->st_uid),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("gid", st->st_gid),
+                        SD_JSON_BUILD_PAIR_CONDITION(
+                                        st->st_rdev != 0,
+                                        "rdev",
+                                        SD_JSON_BUILD_OBJECT(
+                                                        SD_JSON_BUILD_PAIR_UNSIGNED("major", major(st->st_rdev)),
+                                                        SD_JSON_BUILD_PAIR_UNSIGNED("minor", minor(st->st_rdev)))),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("size", st->st_size),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("blockSize", st->st_blksize),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("blocks", st->st_blocks));
+}
+
+int json_variant_new_fd(sd_json_variant **ret, int fd) {
```

Comment:
maybe call this …_new_fd_info()? after all this does not really serialize the fd concept itself (it couldn't, there is no side channel here), but mostly descriptive stuff over the fd.